Purchase shipping labels p1

hichamboushaba opened the pull request:
First part of #3698, it implements shipping labels purchase functionality, unit tests will follow in an upcoming PR.

Please note that the commit 671a0f0 is not related to the feature, it's just an improvement to avoid casting from Double to Flow when calling FluxC's methods.

### Screenshots:
<img width=400 src="https://user-images.githubusercontent.com/1657201/113137213-273d4b00-921c-11eb-8d95-b28a07cb9955.png"/> <img width=400 src="https://user-images.githubusercontent.com/1657201/113137226-299fa500-921c-11eb-9a4f-59bf6ae8850c.png"/>
<img width=400 src="https://user-images.githubusercontent.com/1657201/113141404-94071400-9221-11eb-9bb9-a2297da774ec.png"/> <img width=400 src="https://user-images.githubusercontent.com/1657201/113320370-5a5e0800-930a-11eb-85c5-b0be55b8af3b.png" />

### Question
When checking the option "Mark order as complete", we will make another API call to update the order status, to avoid re-purchasing the label if this call fails, I'm just showing an error SnackBar, then continue to the print label screen, WDYT about that @woocommerce/mobile-designers?
<img width=360 src="https://user-images.githubusercontent.com/1657201/113291286-b6189900-92ea-11eb-994a-0a5229dd1ca0.png" />


### Testing:
1. As this needs some API changes that are not released yet, you need to update WCShip using this [file](https://github.com/wordpress-mobile/WordPress-FluxC-Android/files/6210751/woocommerce-services.zip) in your store.
2. Create an order that satisfies shipping label creation.
3. Go to the order detail, and click on Create shipping label.
4. Pass all the steps.
5. Click on Purchase shipping label.
6. Notice that a loading dialog appears.
7. Confirm that the success screen is shown, and that you can print the label.
8. Repeat the steps from 2 to 4.
9. Check the option "Mark order as complete"
10. Click on  Purchase shipping label.
11. Confirm that after going back to order detail, the order's status has been updated to "completed".

PS: I noticed an issue in the label sizes description screen, I'll create a separate issue to check it.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hichamboushaba committed:
Change the shipping label dimensions to float

The fluxc's implementation uses Float, so we can avoid some casting by making this change

hichamboushaba committed:
Add implementation of shipping label purchase

hichamboushaba committed:
Handle the purchase flow

hichamboushaba committed:
Show the print labels screen after purchase

hichamboushaba committed:
Update print fragment to handle first printing and reprinting cases

hichamboushaba committed:
Update the order detail screen after a label purchase

hichamboushaba committed:
Update print label screen title

hichamboushaba committed:
Fix ktlint issues

hichamboushaba committed:
Handle order fulfilling after purchase

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt:
```diff
@@ -332,6 +355,32 @@ class CreateShippingLabelViewModel @AssistedInject constructor(
+                    remoteOrderId = orderIdSet.remoteOrderId,
+                    newStatus = CoreOrderStatus.COMPLETED.value
+                )
+                if (!fulfillResult) {
```
If the call failed, we just show a SnackBar to let the user knows, but we continue to next screen, to avoid re-purchasing the labels.

hichamboushaba committed:
Merge remote-tracking branch 'origin/develop' into issue/3698-purchase-shipping-labels

# Conflicts:
#	build.gradle

hichamboushaba committed:
Use correct wording for fulfilling failure

0nko commented:
Seeing the entire flow work is amazing! ❤️ 

When I created a SL and then try to refund it, I'm getting `Error: The WooCommerce Shipping & Tax server returned: Bad Request Unable to request refund. The parcel has been shipped. ( 400 )`. I'm just wondering if this is something that we've seen before or it's related to the SL being created on mobile. cc: @anitaa1990

hichamboushaba commented:
Thanks @0nko, regarding refunding labels, in staging you have to do it within 2 minutes of creation (source: p91TBi-3xD-p2#refund-shipping-labels), on my side I can refund labels with this condition, can you check on your side?

0nko commented:
Yup, that worked 🥳

0nko commented:
I noticed that the order detail isn't refreshed after a refund. Maybe we could add that.

hichamboushaba commented:
Yes, I noticed it too, but forgot to mention it, I prefer to create a separate issue to fix it later, WDYT?

hichamboushaba commented:
@0nko FYI: https://github.com/woocommerce/woocommerce-android/issues/3788

0nko reviewed (approved):
Nice work! 🎖️

hichamboushaba committed:
Update the illustration for dark mode

hichamboushaba commented:
I just noticed that the illustration in the print screen is different in dark theme from light one, I updated it, and updated the screenshot in the PR description.

hichamboushaba committed:
Merge remote-tracking branch 'origin/develop' into issue/3698-purchase-shipping-labels

# Conflicts:
#	build.gradle

Garance91540 commented:
Hi @hichamboushaba 

> When checking the option "Mark order as complete", we will make another API call to update the order status, to avoid re-purchasing the label if this call fails, I'm just showing an error SnackBar, then continue to the print label screen, WDYT about that @woocommerce/mobile-designers?

Sounds good!
